src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,335 @@
+        else if (code == "FunctionSyntaxError")
+            return ServiceErrorCode::function_syntax_error;
+        else if (code == "FunctionInvalid")
+            return ServiceErrorCode::function_invalid;
+        else if (code == "IncomingWebhookNotFound")
+            return ServiceErrorCode::incoming_webhook_not_found;
+        else if (code == "IncomingWebhookAlreadyExists")
+            return ServiceErrorCode::incoming_webhook_already_exists;
+        else if (code == "IncomingWebhookDuplicateName")
+            return ServiceErrorCode::incoming_webhook_duplicate_name;
+        else if (code == "RuleNotFound")
+            return ServiceErrorCode::rule_not_found;
+        else if (code == "APIKeyNotFound")
+            return ServiceErrorCode::api_key_not_found;
+        else if (code == "RuleAlreadyExists")
+            return ServiceErrorCode::rule_already_exists;
+        else if (code == "AuthProviderDuplicateName")
+            return ServiceErrorCode::auth_provider_duplicate_name;
+        else if (code == "RestrictedHost")
+            return ServiceErrorCode::restricted_host;
+        else if (code == "APIKeyAlreadyExists")
+            return ServiceErrorCode::api_key_already_exists;
+        else if (code == "IncomingWebhookAuthFailed")
+            return ServiceErrorCode::incoming_webhook_auth_failed;
+        else if (code == "ExecutionTimeLimitExceeded")
+            return ServiceErrorCode::execution_time_limit_exceeded;
+        else if (code == "NotCallable")
+            return ServiceErrorCode::not_callable;
+        else if (code == "UserAlreadyConfirmed")
+            return ServiceErrorCode::user_already_confirmed;
+        else if (code == "UserNotFound")
+            return ServiceErrorCode::user_not_found;
+        else if (code == "UserDisabled")
+            return ServiceErrorCode::user_disabled;
+        else
+            return ServiceErrorCode::unknown;
+    }
+
+    const ServiceErrorCode m_code;
+    const std::string m_raw_code;
+};
+
+} // namespace error
+
+
+enum class Method {
+    get, post, patch
+};
+
+/**
+ * An HTTP request that can be made to an arbitrary server.
+ */
+struct Request {
+    /**
+     * The HTTP method of this request.
+     */
+    Method method;
+
+    /**
+     * The URL to which this request will be made.
+     */
+    std::string url;
+
+    /**
+     * The number of milliseconds that the underlying transport should spend on an HTTP round trip before failing with an
+     * error.
+     */
+    uint64_t timeout_ms;
+
+    /**
+     * The HTTP headers of this request.
+     */
+    std::map<std::string, std::string> headers;
+
+    /**
+     * The body of the request.
+     */
+    std::string body;
+};
+
+/**
+ * The contents of an HTTP response.
+ */
+struct Response {
+    /**
+     * The status code of the HTTP response.
+     */
+    int http_status_code;
+
+    /**
+     * A custom status code provided by the language binding.
+     */
+    int binding_status_code;
```

Comment:
We should probably clarify in docs that this `value != 0` indicates an SDK error? Or at least that is the semantics I had in mind?

Reply:
👍